pkg/module/go_get_fetcher.go

```diff
@@ -0,0 +1,169 @@
+	if !isVgoInstalled() {
+		return nil, errors.New("vgo not installed")
+	}
+
+	if repoURI == "" {
+		return nil, errors.New("invalid repository identifier")
+	}
+
+	return &genericFetcher{
+		fs:      fs,
+		repoURI: repoURI,
+		version: version,
+	}, nil
+}
+
+// Fetch downloads the sources and returns path where it can be found
+func (g *genericFetcher) Fetch(mod, ver string) (Ref, error) {
+	escapedURI := strings.Replace(g.repoURI, "/", "-", -1)
+	repoDirName := fmt.Sprintf(tmpRepoDir, escapedURI, g.version)
+
+	gopath, repoRoot, err := setupTmp(g.fs, repoDirName)
+	if err != nil {
+		return nil, err
+	}
+	g.dirName = repoRoot
+
+	prepareStructure(g.fs, repoRoot)
+
+	dirName, err := getSources(g.fs, gopath, repoRoot, g.repoURI, g.version)
+	diskRef := newDiskRef(g.fs, dirName, ver)
+
+	return diskRef, err
+}
+
+// Clear removes all downloaded data
+func (g *genericFetcher) Clear() error {
+	if g.dirName == "" {
+		return nil
+	}
+
+	return g.fs.RemoveAll(g.dirName)
+}
+
+func isVgoInstalled() bool {
+	cmd := exec.Command("vgo")
+	err := cmd.Run() // this will fail even if vgo is installed
+
+	if err != nil {
+		_, ok := err.(*exec.ExitError) // if vgo is installed it will return ExitError, otherwise normal error
+		return ok == true
+	}
+
+	return false
+}
+
+func setupTmp(fs afero.Fs, repoDirName string) (string, string, error) {
+	gopathDir, err := afero.TempDir(fs, "", "")
+	if err != nil {
+		return "", "", err
+	}
+
+	path := filepath.Join(gopathDir, "src", repoDirName)
+
+	return gopathDir, path, fs.MkdirAll(path, os.ModeDir|os.ModePerm)
+}
+
+// Hacky thing makes vgo not to complain
+func prepareStructure(fs afero.Fs, repoRoot string) error {
+	// vgo expects go.mod file present with module statement or .go file with import comment
+	gomodPath := filepath.Join(repoRoot, "go.mod")
+	gomodContent := []byte("module \"mod\"")
+	if err := afero.WriteFile(fs, gomodPath, gomodContent, 0666); err != nil {
+		return err
+	}
+
+	sourcePath := filepath.Join(repoRoot, "mod.go")
+	sourceContent := []byte(`package mod // import "mod"`)
+	return afero.WriteFile(fs, sourcePath, sourceContent, 0666)
+}
+
+func getSources(fs afero.Fs, gopath, repoRoot, repoURI, version string) (string, error) {
+	version = strings.TrimPrefix(version, "@")
+	if !strings.HasPrefix(version, "v") {
+		version = "v" + version
+	}
+	uri := strings.TrimSuffix(repoURI, "/")
+
+	fullURI := fmt.Sprintf("%s@%s", uri, version)
+
+	gopathEnv := fmt.Sprintf("GOPATH=%s", gopath)
+	cacheEnv := fmt.Sprintf("GOCACHE=%s", filepath.Join(gopath, "cache"))
+	disableCgo := "CGO_ENABLED=0"
+
+	cmd := exec.Command("vgo", "get", fullURI)
+	// PATH is needed for vgo to recognize vcs binaries
+	// this breaks windows.
+	cmd.Env = []string{"PATH=" + os.Getenv("PATH"), gopathEnv, cacheEnv, disableCgo}
+	cmd.Dir = repoRoot
+
+	packagePath := filepath.Join(gopath, "src", "mod", "cache", "download", repoURI, "@v")
+
+	o, err := cmd.CombinedOutput()
```

Comment:
I'm wondering if instead of the error check and eventual return, we can just have the switch.

Since the default case is the same as the return, and if the cmd output contains the limit error we return our custom error.
The only other check is on the files existing, in which case the only time the default case will be executed is when there is some actual error from the vgo get command.

Reply:
I think you don't want to do these checks (isLimit and checkFiles) unnecessarily.
The alternative would be to do something like 
`case err != nil && isLimit`
`case err != nil && checkFiles`
but I'm not sure it's that much better.

Reply:
I've simplified this to not do the switch inside the if. hopefully it's a bit more clear what's going on, even if it's more nested